crypto/ec/ec_mult.c

```diff
@@ -108,61 +108,76 @@ void EC_ec_pre_comp_free(EC_PRE_COMP *pre)
  *
  * It performs either a fixed point multiplication
  *          (scalar * generator)
  * when point is NULL, or a variable point multiplication
  *          (scalar * point)
  * when point is not NULL.
  *
- * scalar should be in the range [0,n) otherwise all constant time bets are off.
+ * scalar cannot be NULL and should be in the range [0,n) otherwise all constant
+ * time bets are off (where n is the cardinality of the EC group).
  *
- * NB: This says nothing about EC_POINT_add and EC_POINT_dbl,
- * which of course are not constant time themselves.
+ * NB: This says nothing about the constant-timeness of the ladder step
+ * implementation (i.e., the default implementation is based on EC_POINT_add and
+ * EC_POINT_dbl, which of course are not constant time themselves) or the
+ * underlying multiprecision arithmetic.
  *
  * The product is stored in r.
  *
+ * This is an internal function: callers are in charge of ensuring that the
+ * input parameters `group`, `r`, `scalar` and `ctx` are not NULL.
+ *
  * Returns 1 on success, 0 otherwise.
  */
-static int ec_mul_consttime(const EC_GROUP *group, EC_POINT *r,
-                            const BIGNUM *scalar, const EC_POINT *point,
-                            BN_CTX *ctx)
+int ec_mul_consttime(const EC_GROUP *group, EC_POINT *r,
+                     const BIGNUM *scalar, const EC_POINT *point,
+                     BN_CTX *ctx)
 {
     int i, cardinality_bits, group_top, kbit, pbit, Z_is_one;
+    EC_POINT *p = NULL;
     EC_POINT *s = NULL;
     BIGNUM *k = NULL;
     BIGNUM *lambda = NULL;
     BIGNUM *cardinality = NULL;
-    BN_CTX *new_ctx = NULL;
     int ret = 0;
 
-    if (ctx == NULL && (ctx = new_ctx = BN_CTX_secure_new()) == NULL)
-        return 0;
+    /* early exit if the input point is the point at infinity */
+    if (point != NULL && EC_POINT_is_at_infinity(group, point))
+        return EC_POINT_set_to_infinity(group, r);
 
     BN_CTX_start(ctx);
 
-    s = EC_POINT_new(group);
-    if (s == NULL)
+    if (((p = EC_POINT_new(group)) == NULL)
+        || ((s = EC_POINT_new(group)) == NULL)) {
+        ECerr(EC_F_EC_MUL_CONSTTIME, ERR_R_MALLOC_FAILURE);
         goto err;
+    }
 
     if (point == NULL) {
-        if (!EC_POINT_copy(s, group->generator))
+        if (!EC_POINT_copy(p, group->generator))
             goto err;
     } else {
-        if (!EC_POINT_copy(s, point))
+        if (!EC_POINT_copy(p, point))
             goto err;
     }
 
+    EC_POINT_BN_set_flags(p, BN_FLG_CONSTTIME);
+    EC_POINT_BN_set_flags(r, BN_FLG_CONSTTIME);
     EC_POINT_BN_set_flags(s, BN_FLG_CONSTTIME);
 
     cardinality = BN_CTX_get(ctx);
     lambda = BN_CTX_get(ctx);
     k = BN_CTX_get(ctx);
-    if (k == NULL || !BN_mul(cardinality, group->order, group->cofactor, ctx))
+    if (k == NULL) {
+        ECerr(EC_F_EC_MUL_CONSTTIME, ERR_R_MALLOC_FAILURE);
+        goto err;
+    }
+
+    if (!BN_mul(cardinality, group->order, group->cofactor, ctx))
         goto err;
```

Comment:
There seems to be inconsistency in this function about whether we call `ECerr` or not.

Reply:
The "fixup!" commits below fix the reported formatting nits in the corresponding commits.

I will have now a look at `ec_mul_consttime` to raise `ECerr` in the appropriate cases.

Reply:
@mattcaswell can you check 6f9fd585222 ?

(I couldn't find more specific reasons than `ERR_R_EC_LIB`/`ERR_R_BN_LIB`). Suggestions?